client/relay-chain-local/src/lib.rs

```diff
@@ -0,0 +1,542 @@
+		network.is_major_syncing()
+	}
+
+	fn overseer_handle(&self) -> Option<Handle> {
+		self.overseer_handle.clone()
+	}
+
+	fn get_storage_by_key(
+		&self,
+		block_id: &BlockId,
+		key: &[u8],
+	) -> Result<Option<StorageValue>, sp_blockchain::Error> {
+		let state = self.backend.state_at(*block_id)?;
+		state.storage(key).map_err(sp_blockchain::Error::Storage)
+	}
+
+	fn prove_read(
+		&self,
+		block_id: &BlockId,
+		relevant_keys: &Vec<Vec<u8>>,
+	) -> Result<Option<StorageProof>, Box<dyn sp_state_machine::Error>> {
+		let state_backend = self
+			.backend
+			.state_at(*block_id)
+			.map_err(|e| {
+				tracing::error!(
+					target: LOG_TARGET,
+					relay_parent = ?block_id,
+					error = ?e,
+					"Cannot obtain the state of the relay chain.",
+				);
+			})
+			.ok();
+
+		match state_backend {
+			Some(state) => sp_state_machine::prove_read(state, relevant_keys)
+				.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						relay_parent = ?block_id,
+						error = ?e,
+						"Failed to collect required relay chain state storage proof.",
+					);
+					e
+				})
+				.map(Some),
+			None => Ok(None),
+		}
+	}
+
+	/// Wait for a given relay chain block in an async way.
+	///
+	/// The caller needs to pass the hash of a block it waits for and the function will return when the
+	/// block is available or an error occurred.
+	///
+	/// The waiting for the block is implemented as follows:
+	///
+	/// 1. Get a read lock on the import lock from the backend.
+	///
+	/// 2. Check if the block is already imported. If yes, return from the function.
+	///
+	/// 3. If the block isn't imported yet, add an import notification listener.
+	///
+	/// 4. Poll the import notification listener until the block is imported or the timeout is fired.
+	///
+	/// The timeout is set to 6 seconds. This should be enough time to import the block in the current
+	/// round and if not, the new round of the relay chain already started anyway.
+	async fn wait_for_block(&self, hash: PHash) -> Result<(), WaitError> {
+		let mut listener =
+			match check_block_in_chain(self.backend.clone(), self.full_client.clone(), hash)? {
+				BlockCheckStatus::InChain => return Ok(()),
+				BlockCheckStatus::Unknown(listener) => listener,
+			};
+
+		let mut timeout = futures_timer::Delay::new(Duration::from_secs(TIMEOUT_IN_SECONDS)).fuse();
+
+		loop {
+			futures::select! {
+				_ = timeout => return Err(WaitError::Timeout(hash)),
+				evt = listener.next() => match evt {
+					Some(evt) if evt.hash == hash => return Ok(()),
+					// Not the event we waited on.
+					Some(_) => continue,
+					None => return Err(WaitError::ImportListenerClosed(hash)),
+				}
+			}
+		}
+	}
+}
+
+pub enum BlockCheckStatus {
+	/// Block is in chain
+	InChain,
+	/// Block status is unknown, listener can be used to wait for notification
+	Unknown(ImportNotifications<PBlock>),
+}
+
+// Helper function to check if a block is in chain.
+pub fn check_block_in_chain<Client>(
```

Comment:
This helper function came back to solve an issue which the lock guard that was not `Send` in the async function. Also makes this reusable in the test.